In [1]:
import yfinance as yf
import warnings
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

In [2]:
stocks_names=['BHARTIARTL.NS',"SBIN.NS","DRREDDY.NS","GSPL.NS","JSWSTEEL.NS","JSWENERGY.BO","RELIANCE.NS","YESBANK.NS","TATAMOTORS.NS",
"LAURUSLABS.NS",
"NTPC.NS",
"SAIL.BO"]# List of all the stokes names 

In [3]:
stocks=[]
for i in stocks_names:
    data=yf.download(i, start="2015-01-01", end="2022-12-31")
    data=data.reset_index()
    data['stocks']=i
    stocks.append(data)
merged_df = pd.concat(stocks, ignore_index=True) 
# Importing all the data from yfinance and merging all the data in one data set 

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed


In [4]:
merged_df=merged_df.drop(columns=['High','Low','Adj Close','Volume']) # Droping all the unwanted columns in the data set 

In [5]:
merged_df['percentage']=(merged_df['Close']-merged_df['Open'])/merged_df['Open'] 
# finding the percentage of the profit 

In [6]:
final_df=merged_df[merged_df['percentage']>0.053]
# filtering the stocks which has more of profit persentage.

In [7]:
final_df=final_df.sort_values('percentage',ascending=False)

In [8]:
final_df['Date']=final_df['Date'].drop_duplicates()
final_df.dropna(inplace=True)
#Droping all the duplicates from the dataset

In [9]:
final_df=final_df.sort_values('Date')
#sorting the dataset in the decending order

In [10]:
final_df.shape

(294, 5)

In [11]:
final_df

Date        Open       Close         stocks  percentage
5930  2015-01-05  112.699997  122.750000        GSPL.NS    0.089175
2006  2015-02-13  288.000000  306.950012        SBIN.NS    0.065799
7937  2015-02-19   97.879997  103.760002    JSWSTEEL.NS    0.060074
45    2015-03-10  314.062897  337.843719  BHARTIARTL.NS    0.075720
5982  2015-03-23  116.750000  123.400002        GSPL.NS    0.056959
...          ...         ...         ...            ...         ...
7878  2022-11-25  264.000000  278.799988        GSPL.NS    0.056061
15792 2022-12-09   17.700001   19.700001     YESBANK.NS    0.112994
15793 2022-12-12   19.950001   21.100000     YESBANK.NS    0.057644
15794 2022-12-13   21.500000   23.950001     YESBANK.NS    0.113954
15803 2022-12-26   17.600000   19.600000     YESBANK.NS    0.113636

[294 rows x 5 columns]

In [12]:
excel_file = pd.DataFrame(columns=['Dates', 'Name_of_stocks', 'Transaction price','Event(Sell/Buy)','Total Units'])

In [13]:
best_entry = None # declaring the entry point
best_exit = None # declaring the exit point
wallet=100000 # inisial price for trading
list_transaction=[]# declaring the list for collecting transaction price
unit_num=[] # declaring the list for collecting number of unit 
for i in range(len(final_df)): # loop for making the transaction
            entry_price = final_df.iloc[i]["Open"] #initialization of entry point
            exit_price = final_df.iloc[i]["Close"] #initialization of exit point
            units = wallet // entry_price # Calculate number of units that can be bought with initial amount          
            # Calculate profit for the current entry and exit points
            transaction_cost = 0.05 * exit_price * units # Calculate transaction cost for selling stocks
            profit = (exit_price * units)- (entry_price * units) - transaction_cost #profit for every entry and exit
            wallet=(exit_price*units)-transaction_cost # initialization of wallet
            row=wallet
            list_transaction.append(row) #creating the list of transaction amount
            unit_num.append(units) # creating the list of number of units
print('Final amount',wallet) # printing the amount


Final amount 224907744.49732113


### From the above output we can see that maximum amount is 224907744.49732113 

In [14]:
final_df['num_units']=unit_num # appending the number of units to the dataframe

In [15]:
final_df['amount']=list_transaction  # appending the transaction amount to the dataframe

In [16]:
final_df # the final dataframe

Date        Open       Close         stocks  percentage  \
5930  2015-01-05  112.699997  122.750000        GSPL.NS    0.089175   
2006  2015-02-13  288.000000  306.950012        SBIN.NS    0.065799   
7937  2015-02-19   97.879997  103.760002    JSWSTEEL.NS    0.060074   
45    2015-03-10  314.062897  337.843719  BHARTIARTL.NS    0.075720   
5982  2015-03-23  116.750000  123.400002        GSPL.NS    0.056959   
...          ...         ...         ...            ...         ...   
7878  2022-11-25  264.000000  278.799988        GSPL.NS    0.056061   
15792 2022-12-09   17.700001   19.700001     YESBANK.NS    0.112994   
15793 2022-12-12   19.950001   21.100000     YESBANK.NS    0.057644   
15794 2022-12-13   21.500000   23.950001     YESBANK.NS    0.113954   
15803 2022-12-26   17.600000   19.600000     YESBANK.NS    0.113636   

        num_units        amount  
5930        887.0  1.034353e+05  
2006        359.0  1.046853e+05  
7937       1069.0  1.053735e+05  
45          335.0  1.075188e+05  
5982        920.0  1.078516e+05  
...           ...           ...  
7878     713922.0  1.890894e+08  
15792  10683014.0  1.999326e+08  
15793  10021684.0  2.008847e+08  
15794   9343472.0  2.125874e+08  
15803  12078826.0  2.249077e+08  

[294 rows x 7 columns]

In [17]:
#final_df.to_csv(r"C:\Users\LokeshRamesh\Documents\co_10 training\coding challenge\week2\margin_final_101.csv")

In [18]:
out_file=pd.DataFrame() #creating the empty dataframe for importing

In [19]:
#appending all the requied columns to data frame
out_file['Date']=final_df['Date']
out_file['Name of the stocks']=final_df['stocks']
out_file['Transaction_Price_open']=final_df['Open']
out_file['Transaction_Price_close']=final_df['Close']
out_file['Units']=final_df['num_units']

In [20]:
out_file

Date Name of the stocks  Transaction_Price_open  \
5930  2015-01-05            GSPL.NS              112.699997   
2006  2015-02-13            SBIN.NS              288.000000   
7937  2015-02-19        JSWSTEEL.NS               97.879997   
45    2015-03-10      BHARTIARTL.NS              314.062897   
5982  2015-03-23            GSPL.NS              116.750000   
...          ...                ...                     ...   
7878  2022-11-25            GSPL.NS              264.000000   
15792 2022-12-09         YESBANK.NS               17.700001   
15793 2022-12-12         YESBANK.NS               19.950001   
15794 2022-12-13         YESBANK.NS               21.500000   
15803 2022-12-26         YESBANK.NS               17.600000   

       Transaction_Price_close       Units  
5930                122.750000       887.0  
2006                306.950012       359.0  
7937                103.760002      1069.0  
45                  337.843719       335.0  
5982                123.400002       920.0  
...                        ...         ...  
7878                278.799988    713922.0  
15792                19.700001  10683014.0  
15793                21.100000  10021684.0  
15794                23.950001   9343472.0  
15803                19.600000  12078826.0  

[294 rows x 5 columns]

In [21]:
excel=out_file.melt(id_vars=['Units','Name of the stocks','Date']).sort_values(by='Date') 
# creating all the requied columns for the dataframe

In [22]:
excel['Transaction price']=excel['variable'].map(lambda x:x.split('_')[2])

In [23]:
excel.drop(columns='variable',inplace=True)


In [24]:
excel# the final Dataframe for download

Units Name of the stocks       Date       value Transaction price
0         887.0            GSPL.NS 2015-01-05  112.699997              open
294       887.0            GSPL.NS 2015-01-05  122.750000             close
295       359.0            SBIN.NS 2015-02-13  306.950012             close
1         359.0            SBIN.NS 2015-02-13  288.000000              open
2        1069.0        JSWSTEEL.NS 2015-02-19   97.879997              open
..          ...                ...        ...         ...               ...
291  10021684.0         YESBANK.NS 2022-12-12   19.950001              open
586   9343472.0         YESBANK.NS 2022-12-13   23.950001             close
292   9343472.0         YESBANK.NS 2022-12-13   21.500000              open
293  12078826.0         YESBANK.NS 2022-12-26   17.600000              open
587  12078826.0         YESBANK.NS 2022-12-26   19.600000             close

[588 rows x 5 columns]

In [25]:
excel.to_excel(r'C:\Users\LokeshRamesh\Documents\co_10 training\coding challenge\week2\coding_challeng_week2.xlsx')